elastic/kibana
# [Synthetics] fix alert actions for synthetics monitor status rule

## Summary

Resolves https://github.com/elastic/kibana/issues/151796
Resolves https://github.com/elastic/kibana/issues/151795

The index alert connector is shared by both legacy uptime and synthetics, but the context shape is different between the two alert types. To account for this a `isLegacy` boolean was added that determines which index connector configuration is used. 

Additionally, the recovery message was missing when the monitor resolves due to status up. This PR ensures that a recovery message is still populated when the monitor recovers after coming up.

Example server log for down monitor
```
[2023-02-22T13:46:04.521-05:00][WARN ][plugins.actions.server-log] Server log: The monitor Badssl checking https://expired.badssl.com/ from US Central QA last ran at February 22, 2023 1:45 PM and is down. The last error received is: error executing step: page.goto: net::ERR_CERT_DATE_INVALID at https://expired.badssl.com/.
```
Example slack message for down monitor
<img width="1428" alt="Screen Shot 2023-02-22 at 1 47 57 PM" src="https://user-images.githubusercontent.com/11356435/220729032-a15341cd-3123-4c28-a32e-e98ccf5dfaa8.png">

Example 

Example server log for recovery
```
[2023-02-22T13:16:45.058-05:00][INFO ][plugins.actions.server-log] Server log: The alert for the monitor Badssl checking https://expired.badssl.com/ from North America - US Central is no longer active: Monitor has recovered with status Up.
```

Example slack message for recovery
<img width="1148" alt="Screen Shot 2023-02-22 at 1 18 40 PM" src="https://user-images.githubusercontent.com/11356435/220720380-8342af47-5de6-49fe-a58b-5bf4df460093.png">

Example index document for recovery
```
{
        "_index": "test3",
        "_id": "WnZieoYBn2Xs-NzyH0X7",
        "_score": 1,
        "_source": {
          "monitorName": "https://elastic.co",
          "monitorUrl": "https://elastic.co/",
          "statusMessage": "up",
          "latestErrorMessage": "",
          "observerLocation": "North America - US Central",
          "recoveryReason": "Monitor has recovered with status Up"
        }
      }
```

```
{
        "_index": "test3",
        "_id": "z1RseoYBfB3Am9NRsxsx",
        "_score": 1,
        "_source": {
          "monitorName": "Beep",
          "monitorUrl": "about:blank",
          "statusMessage": "down",
          "latestErrorMessage": "",
          "observerLocation": "North America - US Central",
          "recoveryReason": "Location has been removed from the monitor"
        }
      }
```

```
{
        "_index": "test3",
        "_id": "rVRreoYBfB3Am9NRvRsu",
        "_score": 1,
        "_source": {
          "monitorName": "TCP monitor",
          "monitorUrl": "tcp://localhost:5601",
          "statusMessage": "down",
          "latestErrorMessage": "",
          "observerLocation": "North America - US Central",
          "recoveryReason": "Monitor has been deleted"
        }
      },
```

Testing
--

1. Create a default alert connector on the Synthetics settings page. This issue in particular impacted the `index` connector, so it's recommended to create an index connector. For smoke testing, it's also good to create an server log connector, or a slack connector.
2. Create an always down monitor, and ideally run it against two locations (if using the dev service environment, you can use US Central dev and US Central QA). 
3. Wait for the alert to fire
4. Confirm that all your alert connectors are working with all content populated. In particular, query against the index that you are using for the index connector. Confirm that all the fields are populated appropriately in the document.
5. Remove one of the locations from your monitor
6. Wait for the alert for that location to resolve. Confirm that the recovery message for that location was sent to all your connectors. In particular, query against the index that you are using for the index connector. Find the recovery document. Ensure that all the fields are populated appropriately
7. Update the monitor so it now fires as up
8. Wait for the alert to resolve. Confirm that the recovery message for the remaining location was sent to all your alert connectors. In particular, query against the index that you are using for the index connector. Find the recovery document. Ensure that the reason message mentions that the monitor is now up and the status is marked as up.
9. Update the monitor so that it becomes down again
10. Wait for the alert to fire
11. After the alert fires for the down monitor, delete the monitor
12. Wait for the alert to resolve. Confirm that the recovery message was sent to all your alert connectors stating that the alert was resolved because the monitor was deleted. In particular, query against the index that you are using for the index connector. Find the recovery document. Ensure that the reason message mentions that the monitor was deleted.
